storage/innobase/xtrabackup/test/t/xbcloud.sh

```diff
@@ -165,8 +169,111 @@ if  grep -q "Upload failed" $topdir/pxb-2279.log ; then
     die 'xbcloud exit with error on upload'
 fi
 
+# PXB-2112 - Support for storage class
+test_storage_class=false
+
+if grep -q "storage=s3" $topdir/xbcloud.cnf ; then
+  vlog "Testing S3 storage class"
+  test_storage_class=true
+  storage_class_parameter="--s3-storage-class=GLACIER"
+elif grep -q "storage=google" $topdir/xbcloud.cnf ; then
+  vlog "Testing GC storage class"
+  test_storage_class=true
+  storage_class_parameter="--google-storage-class=COLDLINE"
+fi
+
+if grep -q "minioadmin" $topdir/xbcloud.cnf ; then
+  test_storage_class=false
+fi
+
+if [ "$test_storage_class" = true ]; then
+  run_cmd touch $topdir/xtrabackup_info
+  cd $topdir/
+  xbstream -c xtrabackup_info | \
+  xbcloud --defaults-file=$topdir/xbcloud.cnf ${storage_class_parameter} \
+   put ${storage_class_folder} 2> $topdir/storage_class.log
+
+ if  grep -q "expected objects using storage class" $topdir/storage_class.log ; then
+    xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
+     ${storage_class_folder}
+     die 'xbcloud did not upload using correct storage class'
+ fi
+ if  grep -q "Upload failed" $topdir/storage_class.log ; then
+   xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
+     ${storage_class_folder}
+     die 'xbcloud exit with error on upload'
+ fi
+ run_cmd xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
+  ${storage_class_folder}
+fi
+
 # cleanup
 run_cmd xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
 	${full_backup_name} --parallel=4
 run_cmd xbcloud --defaults-file=$topdir/xbcloud.cnf delete \
 	${inc_backup_name} --parallel=4
+rm -rf ${full_backup_dir}
+rm -rf $topdir/downloaded_full
+
+
+#check if password is stored in backup directory or in ps output
+
+## grep for all key in secret_key
+params=($(echo $XBCLOUD_CREDENTIALS | tr ";" "\n"))
+secret_keys=();
+for param in "${params[@]}"
+do
+ if [[ $param =~ "key=" ]]; then
+	 secret_keys+=(`echo $param | awk -F 'key=' '{print $2}'`)
+ fi
+done
+
+### create table in MyISAM to let backup wait until there is pending update
+mysql -e "CREATE TABLE t1(i int) engine=MYISAM" test
+mysql -e "INSERT into t1 values(100)" test
+
+### The running update on myisam table will hold access to the table.
+### PXB will do FLUSH TABLES on MyISAM tables which requires access
+### to the table and waits for the running update on MyISAM to finish.
+mysql -e "SELECT SLEEP(1000) from t1 for update " test &
+
+xtrabackup --backup --stream=xbstream --extra-lsndir=$full_backup_dir \
+	   --target-dir=$full_backup_dir | \
+    run_cmd xbcloud $XBCLOUD_CREDENTIALS put \
+	    --parallel=4 \
+	    ${full_backup_name} &
+job_pid=$!
```

Comment:
we are using pipe here, what are we running in background ? pxb + xbcloud or only xbcloud?
You 100% sure that $! give the pid of xbcloud? like for return code we have PIPESTATUS[0] and PIPESTATUS[1] - do we have something similar to catch the PID?

Reply:
both are in background. I tested the test case by commenting AZURE_SECRET_KEY for hide option and test failed. so no worries. 

for example $! is 8068 and below process contains 8068

 ps -ef |grep 8068
  501  8068  7343   0 10:47AM ttys001    0:00.00 /bin/bash ./run.sh -t t/xbcloud.sh
  501  8084  8068   0 10:47AM ttys001    0:00.63 xbcloud --storage=s3 --s3-access-key=
  501  8534  8512   0 10:48AM ttys004    0:00.00 grep 8068

Reply:
Rahul, which pid will be captured in job_pid? Later we do "run_cmd wait $job_pid" so what do we wait for to finish?

Reply:
some new learning for me. If I don't use run_cmd and uses only xbcloud job_pid captures the xbcloud pid.  else It will store pid of main script. so removed the run_cmd.
fixed

Reply:
let me try the backgroup process as well missed that comment will update you.